File: backend/mr/src/main/java/io/featurehub/mr/resources/oauth2/OAuth2MRAdapter.java
package io.featurehub.mr.resources.oauth2;

import cd.connect.app.config.ConfigKey;
import cd.connect.app.config.DeclaredConfigResolver;
import io.featurehub.db.api.AuthenticationApi;
import io.featurehub.db.api.OptimisticLockingException;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.OrganizationApi;
import io.featurehub.db.api.PersonApi;
import io.featurehub.db.services.InternalOAuthPersonCreation;
import io.featurehub.mr.auth.AuthenticationRepository;
import io.featurehub.mr.model.Person;
import io.featurehub.web.security.oauth.SSOCompletionListener;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import jakarta.ws.rs.core.NewCookie;
import jakarta.ws.rs.core.Response;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;

import static jakarta.ws.rs.core.Cookie.DEFAULT_VERSION;
import static jakarta.ws.rs.core.NewCookie.DEFAULT_MAX_AGE;

@Singleton
public class OAuth2MRAdapter implements SSOCompletionListener {
  private static final Logger log = LoggerFactory.getLogger(OAuth2MRAdapter.class);

  protected final PersonApi personApi;
  private final AuthenticationApi authenticationApi;
  private final AuthenticationRepository authRepository;
  private final OrganizationApi organizationApi;
  private final InternalOAuthPersonCreation internalOAuthPersonCreation;

  @ConfigKey("oauth2.cookie.domain")
  String cookieDomain = "";

  @ConfigKey("oauth2.cookie.https-only")
  Boolean cookieSecure = Boolean.FALSE;

  @Inject
  public OAuth2MRAdapter(
      PersonApi personApi,
      AuthenticationApi authenticationApi,
      AuthenticationRepository authRepository,
      OrganizationApi organizationApi,
      InternalOAuthPersonCreation internalOAuthPersonCreation) {
    this.personApi = personApi;
    this.authenticationApi = authenticationApi;
    this.authRepository = authRepository;
    this.organizationApi = organizationApi;
    this.internalOAuthPersonCreation = internalOAuthPersonCreation;

    DeclaredConfigResolver.resolve(this);
  }

  @NotNull
  @Override
  public Response successfulCompletion(
      @Nullable String email,
      @Nullable String username,
      boolean userMustBeCreatedFirst,
      @Nullable String failureUrl,
      @Nullable String successUrl,
      @NotNull String provider) {
    if (email == null || username == null) {
      log.warn("User tried to login via SSO with no email `` or username ``", email, username);

      return Response.status(302).location(URI.create(failureUrl)).build();
    }
    // discover if they are a user and if not, add them
    Person p = personApi.get(email, Opts.empty());

    if (p != null && p.getWhenArchived() != null) {
      log.warn("User {} attempted to login and have been deleted.", email);
      return Response.status(302).location(URI.create(failureUrl)).build();
    }

    if (p == null) {
      // if the user must be created in the database before they are allowed to sign in, redirect to
      // failure.
      if (userMustBeCreatedFirst) {
        log.warn(
            "User {} attempted to login and they aren't in the database and they need to be.",
            email);
        return Response.status(302).location(URI.create(failureUrl)).build();
      }

      p = internalOAuthPersonCreation.createUser(email, username);
    } else {
      p.setName(username);

      try {
        p.setGroups(null); // don't update groups.
        // this is our single write, all in one transaction so we should be good for HA
        personApi.update(p.getId().getId(), p, Opts.empty(), p.getId().getId());
      } catch (OptimisticLockingException ignored) {
      }

      authenticationApi.updateLastAuthenticated(p.getId().getId());
    }

    // store user in session with bearer token
    String token = authRepository.put(p);

    URI uri = URI.create(successUrl);
    // add cookie
    return Response.status(Response.Status.FOUND)
        .cookie(
            new NewCookie(
                "bearer-token",
                token,
                "/",
                cookieDomain.isEmpty() ? null : cookieDomain,
                DEFAULT_VERSION,
                null,
                DEFAULT_MAX_AGE,
                null,
                cookieSecure,
                false))
        .location(uri)
        .build();
  }

  @Override
  public boolean initialAppSetupComplete() {
    return organizationApi.get() != null;
  }
}

File: backend/mr/src/main/kotlin/io/featurehub/mr/resources/oauth2/OAuth2MRAdapter.kt
package io.featurehub.mr.resources.oauth2

import cd.connect.app.config.ConfigKey
import cd.connect.app.config.DeclaredConfigResolver
import io.featurehub.db.api.*
import io.featurehub.mr.auth.AuthenticationRepository
import io.featurehub.mr.model.*
import io.featurehub.mr.utils.PortfolioUtils
import io.featurehub.web.security.oauth.SSOCompletionListener
import jakarta.inject.Inject
import jakarta.inject.Singleton
import jakarta.ws.rs.core.Cookie
import jakarta.ws.rs.core.NewCookie
import jakarta.ws.rs.core.Response
import org.slf4j.LoggerFactory
import java.net.URI

@Singleton
class OAuth2MRAdapter @Inject constructor(
  protected val personApi: PersonApi,
  private val authenticationApi: AuthenticationApi,
  private val portfolioApi: PortfolioApi,
  private val groupApi: GroupApi,
  private val authRepository: AuthenticationRepository,
  private val portfolioUtils: PortfolioUtils,
  private val organizationApi: OrganizationApi
) : SSOCompletionListener {
  @ConfigKey("oauth2.cookie.domain")
  var cookieDomain: String? = ""

  @ConfigKey("oauth2.cookie.https-only")
  var cookieSecure: Boolean? = false

  init {
    DeclaredConfigResolver.resolve(this)
  }

  override fun successfulCompletion(
    email: String, username: String?, userMustBeCreatedFirst: Boolean,
    failureUrl: String, successUrl: String, provider: String
  ): Response {
    // discover if they are a user and if not, add them
    var p = personApi[email, Opts.opts(FillOpts.Archived)]
    if (p != null && p.whenArchived != null) {
      log.warn("User {} attempted to login and have been deleted.", email)
      return Response.status(302).location(URI.create(failureUrl)).build()
    }
    if (p == null) {
      // if the user must be created in the database before they are allowed to sign in, redirect to failure.
      if (userMustBeCreatedFirst) {
        log.warn("User {} attempted to login and they aren't in the database and they need to be.", email)
        return Response.status(302).location(URI.create(failureUrl)).build()
      }
      p = createUser(email, username)
    } else {
      p.name = username

      try {
        personApi.updateV2(p.id!!.id, UpdatePerson().name(username), p.id!!.id)
      } catch (ignored: OptimisticLockingException) {
      }

      authenticationApi.updateLastAuthenticated(p.id!!.id)
    }

    // store user in session with bearer token
    val token = authRepository.put(p)
    val uri = URI.create(successUrl)
    // add cookie
    return Response.status(Response.Status.FOUND).cookie(
      NewCookie.Builder("bearer-token").value(token).path("/")
        .domain(if (cookieDomain!!.isEmpty()) null else cookieDomain)
        .version(Cookie.DEFAULT_VERSION)
        .maxAge(NewCookie.DEFAULT_MAX_AGE)
        .secure(cookieSecure!!)
        .httpOnly(false).build()
    )
      .location(uri).build()
  }

  private fun createUser(email: String?, username: String?): Person? {
    // determine if they were the first user, and if so, complete setup
    val firstUser = personApi.noUsersExist()

    // first we create them, this will give them a token and so forth, we are playing with existing functionality
    // here
    try {
      personApi.create(email!!, username, null)
    } catch (e: PersonApi.DuplicatePersonException) {
      log.error("Shouldn't get here, as we check if the person exists before creating them.")
      return null
    }

    // now "register" them. We can provide a null password OK, it just ignores it, but this removes
    // any registration token required
    val person = authenticationApi.register(username, email, null, null) ?: return null

    if (firstUser) {
      val organization = organizationApi.get()
      // create the superuser group and add admin to the group -
      val group = groupApi.createOrgAdminGroup(organization.id, "org_admin", person)
      groupApi.addPersonToGroup(group!!.id, person.id!!.id, Opts.empty())

      // find the only portfolio and update its members to include this one
      val portfolio = portfolioApi.findPortfolios(
        null, SortOrder.ASC,
        Opts.empty(), person.id!!.id
      )[0]
      try {
        val name = portfolioUtils.formatPortfolioAdminGroupName(portfolio.name)

        groupApi.createGroup(
          portfolio.id,
          CreateGroup().name(name).admin(true), person
        )
      } catch (e: GroupApi.DuplicateGroupException) {
        log.error("If we have this exception, the site is broken.", e)
      }
    }
    return person
  }

  override fun initialAppSetupComplete(): Boolean {
    return organizationApi.get() != null
  }

  companion object {
    private val log = LoggerFactory.getLogger(OAuth2MRAdapter::class.java)
  }
}
